### PR TITLE
Fix #7272 - re-Add modulesPath and add error context on module arguments evaluation.

### LINKED ISSUE #7272: Extra arguments in `configuration.nix` with default values no longer work
After #6794 was merged, my NixOS configuration started to fail with the following error:

```
building the system configuration...
error: while evaluating the attribute ‘config.system.build.toplevel’ at "/nixdev/nixpkgs/nixos/modules/system/activation/top-level.nix":228:5:
(...)
while evaluating the attribute ‘value’ at "/nixdev/nixpkgs/lib/types.nix":119:38:
while evaluating the attribute ‘_module.args."${name}"’ at "/nixdev/nixpkgs/lib/attrsets.nix":135:44:
attribute ‘_module.args.rg_path’ missing, at "/nixdev/nixpkgs/lib/modules.nix":138:17
```

My `configuration.nix` has the following contents:

``` haskell
{ config, pkgs, lib 
, rg_path ? "/root/src/rg"
, ... }:

let
  rgpkg = import "${rg_path}/default.nix" {}; 
in
(...)
  environment.systemPackages = with pkgs; [
    (...)
    rgpkg
  ];
```

I'm guessing it's some issue related to arguments with default values not being handled correctly after the `lib/modules.nix` refactoring?

cc @nbp @shlevy

### PR DESCRIPTION
**Pull request by nbp**
This modification undo the modifications related to `modulesPath` and use it in the installer to ensure that we do not brake this accidentally.  As there is no module logic for the moment, which can be used to replace the `modulesPath` argument, I prefer to add it back in the mean time.

In addition, this patch add an error context which, display one extra line of information when --show-trace is used.

```
while evaluating the module argument `pkgs' in "/home/nicolas/nixos-dev/nixpkgs/foo.nix":
infinite recursion encountered
```

This error message worked for `modulesPath` before this modification which re-enables it, and it also works for the now deprecated use of `with pkgs.lib;`.

@wizeman, can you confirm if this fix your issues?

### CODE REVIEW

**Comment by nbp** [comment]
The second patch add the `importsArgs` feature after all, to filter attributes which should not come from the module system evaluation, even if they are not present in the `extraArgs`.

**Comment by wizeman** [comment]
@nbp I'm not having much success with this fix.

I have done exactly like you said in your April 14th's comment in #7272, with the exception that I've added a `pkgs` argument because I need it in my configuration, e.g.:
```
{ config, pkgs, lib, hydra_path ? "/some/path/hydra", ... }:

{
  imports = [
    ...
    "${hydra_path}/hydra-module.nix"
  ];
  importsArgs = [ "hydra_path" ];
  …
}
```

Even though I've applied these 2 commits, I'm getting this error:
```
building all machine configurations...
error: The option `importsArgs' defined in `/backup/root/src/wizynet/config-homeserv.nix' does not exist.
(use ‘--show-trace’ to show detailed location information)
error: unable to build all machine configurations
```

With `--show-trace`:
```
building all machine configurations...
error: while evaluating the attribute ‘buildCommand’ of the derivation ‘nixops-machines’ at "/nixdev/nixpkgs/pkgs/build-support/trivial-builders.nix":10:14:
while evaluating anonymous function at "/nix/store/dz0v241nyc5bv2x3l5ij54nxfv7m7gab-nixops-1.3pre1455_6ac0752/share/nix/nixops/eval-machine-info.nix":143:46, called from "/nixdev/nixpkgs/lib/attrsets.nix":135:52:
while evaluating the attribute ‘config.system.build.toplevel’ at "/nixdev/nixpkgs/nixos/lib/eval-config.nix":50:5:
while evaluating the attribute ‘config’ at "/nixdev/nixpkgs/lib/modules.nix":82:25:
while evaluating ‘yieldConfig’ at "/nixdev/nixpkgs/lib/modules.nix":69:29, called from "/nixdev/nixpkgs/lib/modules.nix":68:16:
while evaluating ‘fold’ at "/nixdev/nixpkgs/lib/lists.nix":20:19, called from "/nixdev/nixpkgs/lib/modules.nix":75:11:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":27:8:
while evaluating anonymous function at "/nixdev/nixpkgs/lib/modules.nix":75:20, called from "/nixdev/nixpkgs/lib/lists.nix":26:14:
while evaluating ‘fold’ at "/nixdev/nixpkgs/lib/lists.nix":20:19, called from "/nixdev/nixpkgs/lib/modules.nix":76:13:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":27:8:
while evaluating anonymous function at "/nixdev/nixpkgs/lib/modules.nix":76:25, called from "/nixdev/nixpkgs/lib/lists.nix":26:14:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":26:34:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":26:34:
while evaluating anonymous function at "/nixdev/nixpkgs/lib/modules.nix":75:20, called from "/nixdev/nixpkgs/lib/lists.nix":26:14:
while evaluating ‘fold’ at "/nixdev/nixpkgs/lib/lists.nix":20:19, called from "/nixdev/nixpkgs/lib/modules.nix":76:13:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":27:8:
while evaluating anonymous function at "/nixdev/nixpkgs/lib/modules.nix":76:25, called from "/nixdev/nixpkgs/lib/lists.nix":26:14:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":26:34:
while evaluating anonymous function at "/nixdev/nixpkgs/lib/modules.nix":76:25, called from "/nixdev/nixpkgs/lib/lists.nix":26:14:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":26:34:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":26:34:
while evaluating anonymous function at "/nixdev/nixpkgs/lib/modules.nix":75:20, called from "/nixdev/nixpkgs/lib/lists.nix":26:14:
while evaluating ‘fold’ at "/nixdev/nixpkgs/lib/lists.nix":20:19, called from "/nixdev/nixpkgs/lib/modules.nix":76:13:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":27:8:

(...)

while evaluating anonymous function at "/nixdev/nixpkgs/lib/modules.nix":75:20, called from "/nixdev/nixpkgs/lib/lists.nix":26:14:
while evaluating ‘fold’ at "/nixdev/nixpkgs/lib/lists.nix":20:19, called from "/nixdev/nixpkgs/lib/modules.nix":76:13:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":27:8:
while evaluating anonymous function at "/nixdev/nixpkgs/lib/modules.nix":76:25, called from "/nixdev/nixpkgs/lib/lists.nix":26:14:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":26:34:
while evaluating anonymous function at "/nixdev/nixpkgs/lib/modules.nix":76:25, called from "/nixdev/nixpkgs/lib/lists.nix":26:14:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":26:34:
while evaluating anonymous function at "/nixdev/nixpkgs/lib/modules.nix":76:25, called from "/nixdev/nixpkgs/lib/lists.nix":26:14:
while evaluating ‘fold'’ at "/nixdev/nixpkgs/lib/lists.nix":23:15, called from "/nixdev/nixpkgs/lib/lists.nix":26:34:
while evaluating anonymous function at "/nixdev/nixpkgs/lib/modules.nix":76:25, called from "/nixdev/nixpkgs/lib/lists.nix":26:14:
The option `importsArgs' defined in `/backup/root/src/wizynet/config-homeserv.nix' does not exist.
error: unable to build all machine configurations
```

Am I doing something wrong?

**Comment by nbp** [comment]
Can you wrap all option definitions under the config attribute?

```
{ config, pkgs, lib, hydra_path ? "/some/path/hydra", ... }:

{
  imports = [
    ...
    "${hydra_path}/hydra-module.nix"
  ];
  importsArgs = [ "hydra_path" ];
  config = {
    …
  };
}
```

**Comment by wizeman** [comment]
@nbp With your latest suggestion it seems to work fine now. Thanks!

So I guess that putting everything under `config` is needed only if you use `importsArgs`?

**Comment by domenkozar** [comment]
cc @edolstra

**Comment by edolstra** [comment]
Hm, `importArgs` seems pretty ugly to me. I mean, it requires the user to duplicate a function argument that Nix already knows about:
```
{ config, hydraPath, ... }:
{
  importArgs = [ "hydraPath" ];
}
```
Here `hydraPath` already appears in the function header, so it *should* be unnecessary to repeat it.

BTW, why do we need `modulesPath` back?

**Comment by nbp** [comment]
Apparently, since `modulesPath` got removed, a lot of people seems to have uploaded a new version of their own clone of the `amazon-image.nix` file, which unfortunately relies on `modulesPath`, and there is no simple way to remove it.

Yes, `importArgs` is ugly, but for the moment Nix language does not give us a *proper* way to investigate if one argument is necessary or for the evaluation of imports.  (while writing that, I think there might be an ugly way of doing it)

In the mean time, I prefer to add `modulesPath` back, and add `importArgs` attribute, as a work-around until we got a proper module-import logic.

**Comment by nbp** [comment]
If we did not had any issue for introspecting whatever or not an attribute has a default value (see https://github.com/NixOS/nixpkgs/pull/7835#issuecomment-103142770), then we would have had 2 other solutions:

1/ Use `tryEval`, and use a trial and error on the evaluation of the `imports` to determine which set of arguments are required for the evaluation of the imports. Which means that we could have a combinatorial complexity on the resolution of the minimal set of arguments which are required by `imports`.

2/ Have to different values for the module arguments, depending if we are evaluating the `imports`, or if we are evaluating the modules.  If we are evaluating the `imports`, then we would be replace any access to `_module.args` by an assertion, and use the `evalModules` `extraArgs` by default.  If we are evaluating the module, then we would resolve `_module.args` and use the `evalModules` `extraArgs` as forced value of the `_module.args`.

**Comment by edolstra** [comment]
@nbp Uhm, we *have* a way to introspect whether an argument has a default value:
```
$ nix-instantiate --eval -E 'builtins.functionArgs ({ x, y ? 123 }: x)'
{ x = false; y = true; }
```

**Comment by edolstra** [comment]
Ping @nbp. Can you please explain why we need `importArgs`? It strikes me as a really ugly hack that I'd hate to see in a stable release (because then we'll be stuck with it forever).

File: nixos/maintainers/scripts/ec2/amazon-base-config.nix [hunk 1]
```diff
@@ -1,5 +1,5 @@
-{ modulesPath, ...}:
+{ ... }:
 {
-  imports = [ "${modulesPath}/virtualisation/amazon-config.nix" ];
```
Comment (copumpkin) [inline nixos/maintainers/scripts/ec2/amazon-base-config.nix]: This doesn't seem like the right change. I thought you were adding `modulesPath` back in, so why is this needed? Same with azure. Using `<nixpkgs>` there assumes that this file only gets copied into the target filesystem, whereas some image building code also uses it.

**Comment by nbp** [comment]
> Can you please explain why we need `importArgs`? It strikes me as a really ugly hack that I'd hate to see in a stable release (because then we'll be stuck with it forever).

Indeed, `importArgs` is a hack to work-around the fact that we are unable to know if an argument is used to compute the closure, or if it used to compute the content of the modules.  Basically, it serves the same purpose as the `specialArgs` added in #8204, without making a special attribute for it.

I think we should look for a before-closure module logic in a proper way, and I agree that this is an hack for the moment.  I will close this pull request and open a new one for adding error context on the argument evaluation.
